File: csharp/SupermarketReceipt/ReceiptPrinter.cs
using System.Globalization;
using System.Text;

namespace SupermarketReceipt
{
    public class ReceiptPrinter
    {
        private static readonly CultureInfo Culture = CultureInfo.CreateSpecificCulture("en-GB");

        private readonly int _columns;


        public ReceiptPrinter(int columns)
        {
            _columns = columns;
        }

        public ReceiptPrinter() : this(40)
        {
        }

        public string PrintReceipt(Receipt receipt)
        {
            var result = new StringBuilder();
            foreach (var item in receipt.GetItems())
            {
                string receiptItem = PrintReceiptItem(item);
                result.Append(receiptItem);
                
            }

            foreach (var discount in receipt.GetDiscounts())
            {
                string discountPresentation = PrintDiscount(discount);
                result.Append(discountPresentation);
            }

            {
                result.Append("\n");
                result.Append(PrintTotal(receipt));
            }
            return result.ToString();
        }

        private string PrintTotal(Receipt receipt)
        {
            string name = "Total: ";
            string value = PrintPrice(receipt.GetTotalPrice());
            return FormatLineWithWhitespace(name, value);
        }

        private string PrintDiscount(Discount discount)
        {
            string name = discount.Description + "(" + discount.Product.Name + ")";
            string value = PrintPrice(discount.DiscountAmount);

            return FormatLineWithWhitespace(name, value);
        }

        private string PrintReceiptItem(ReceiptItem item)
        {
            string totalPrice = PrintPrice(item.TotalPrice);
            string name = item.Product.Name;
            string line = FormatLineWithWhitespace(name, totalPrice);
            if (item.Quantity != 1)
            {
                line += "  " + PrintPrice(item.Price) + " * " + PrintQuantity(item) + "\n";
            }

            return line;
        }
        

        private string FormatLineWithWhitespace(string name, string value)
        {
            var line = new StringBuilder();
            line.Append(name);
            int whitespaceSize = this._columns - name.Length - value.Length;
            for (int i = 0; i < whitespaceSize; i++) {
                line.Append(" ");
            }
            line.Append(value);
            line.Append('\n');
            return line.ToString();
        }

        private string PrintPrice(double price)
        {
            return price.ToString("N2", Culture);
        }

        private static string PrintQuantity(ReceiptItem item)
        {
            return ProductUnit.Each == item.Product.Unit
                ? ((int) item.Quantity).ToString()
                : item.Quantity.ToString("N3", Culture);
        }
        
    }
}

File: swift/SupermarketReceipt/ReceiptPrinter.swift
public class ReceiptPrinter {

    private var columns: Int = 40

    public init(columns: Int) {
        self.columns = columns
    }

    public func printReceipt(receipt: Receipt) -> String {
        var result = ""
        for item in receipt.items {
            var price = String(format: "%.2f", item.totalPrice)
            var quantity = ReceiptPrinter.presentQuantity(item: item)
            var name = item.product.name
            var unitPrice = String(format :"%.2f", item.price)

            var whitespaceSize = self.columns - name.count - price.count
            var line = name + ReceiptPrinter.getWhitespace(whitespaceSize: whitespaceSize) + price + "\n"

            if (item.quantity != 1) {
                line += "  " + unitPrice + " * " + quantity + "\n"
            }
            result.append(line)
        }
        for discount in receipt.discounts {
            var productPresentation = discount.product.name
            var pricePresentation = String(format: "%.2f", discount.discountAmount)
            var description = discount.description
            result.append(description)
            result.append("(")
            result.append(productPresentation)
            result.append(")")
            result.append(ReceiptPrinter.getWhitespace(whitespaceSize: self.columns - 3 - productPresentation.count - description.count - pricePresentation.count))
            result.append("-")
            result.append(pricePresentation)
            result.append("\n")
        }
        result.append("\n")
        var pricePresentation = String(format: "%.2f", Double(receipt.getTotalPrice()))
        var total = "Total: "
        var whitespace = ReceiptPrinter.getWhitespace(whitespaceSize: self.columns - total.count - pricePresentation.count)
        result.append(total)
        result.append(whitespace)
        result.append(pricePresentation)
        return result
    }

    private static func presentQuantity(item: ReceiptItem ) -> String {
        return ProductUnit.Each == item.product.unit
            ? String(format: "%x", Int(item.quantity))
            : String(format: "%.3f", item.quantity)
    }

    private static func getWhitespace(whitespaceSize: Int) -> String {
        var whitespace = ""
        for i in 0..<whitespaceSize {
            whitespace.append(" ")
        }
        return whitespace
    }
}
